src/apns_supervisor2.erl

```diff
@@ -0,0 +1,1071 @@
+	    {ok, NState};
+	{ok, Pid, _Extra} ->
+	    NState = replace_child(Child#child{pid = Pid}, State),
+	    {ok, NState};
+	{error, Reason} ->
+	    report_error(start_error, Reason, Child, State#state.name),
+	    Restart(Child, State)
+    end;
+restart(rest_for_one, Child, State, Restart) ->
+    {ChAfter, ChBefore} = split_child(Child#child.pid, State#state.children),
+    ChAfter2 = terminate_children(ChAfter, State#state.name),
+    case start_children(ChAfter2, State#state.name) of
+	{ok, ChAfter3} ->
+	    {ok, State#state{children = ChAfter3 ++ ChBefore}};
+	{error, ChAfter3} ->
+	    Restart(Child, State#state{children = ChAfter3 ++ ChBefore})
+    end;
+restart(one_for_all, Child, State, Restart) ->
+    Children1 = del_child(Child#child.pid, State#state.children),
+    Children2 = terminate_children(Children1, State#state.name),
+    case start_children(Children2, State#state.name) of
+	{ok, NChs} ->
+	    {ok, State#state{children = NChs}};
+	{error, NChs} ->
+	    Restart(Child, State#state{children = NChs})
+    end.
+
+%%-----------------------------------------------------------------
+%% Func: terminate_children/2
+%% Args: Children = [#child] in termination order
+%%       SupName = {local, atom()} | {global, atom()} | {pid(),Mod}
+%% Returns: NChildren = [#child] in
+%%          startup order (reversed termination order)
+%%-----------------------------------------------------------------
+terminate_children(Children, SupName) ->
+    terminate_children(Children, SupName, []).
+
+terminate_children([Child | Children], SupName, Res) ->
+    NChild = do_terminate(Child, SupName),
+    terminate_children(Children, SupName, [NChild | Res]);
+terminate_children([], _SupName, Res) ->
+    Res.
+
+terminate_simple_children(Child, Dynamics, SupName) ->
+    Pids = dict:fold(fun (Pid, _Args, Pids) ->
+                         erlang:monitor(process, Pid),
+                         unlink(Pid),
+                         exit(Pid, child_exit_reason(Child)),
+                         [Pid | Pids]
+                     end, [], Dynamics),
+    TimeoutMsg = {timeout, make_ref()},
+    TRef = timeout_start(Child, TimeoutMsg),
+    {Replies, Timedout} =
+        lists:foldl(
+          fun (_Pid, {Replies, Timedout}) ->
+                  {Reply, Timedout1} =
+                      receive
+                          TimeoutMsg ->
+                              Remaining = Pids -- [P || {P, _} <- Replies],
+                              [exit(P, kill) || P <- Remaining],
+                              receive {'DOWN', _MRef, process, Pid, Reason} ->
+                                      {{error, Reason}, true}
+                              end;
+                          {'DOWN', _MRef, process, Pid, Reason} ->
+                              {child_res(Child, Reason, Timedout), Timedout};
+                          {'EXIT', Pid, Reason} ->
+                              receive {'DOWN', _MRef, process, Pid, _} ->
```

Comment:
The expression on line 674 and column 31 is nested beyond the maximum level of 3.